[wasm][wbt] Test that `dotnet.js` could be run from any current directory

yamachu opened the pull request:
In https://github.com/dotnet/runtime/pull/72275 we fixed `locateFile` to deal with relative file path or URL. We could also handle absolute path/URL. This PR is adding tests to prove it works.

pavelsavara commented:
/azp run runtime-wasm

pavelsavara commented:
hi @yamachu, thanks for the contribution!
At first look I like it, it brings more consistency to locating assets.

It would be great to add [unit test](https://github.com/dotnet/runtime/tree/main/src/tests/BuildWasmApps/Wasm.Build.Tests) to cover running V8 and NodeJS from different folder. 
I think we are setting the folder here 
https://github.com/dotnet/runtime/blob/da2cf17084c837d5c14fba173b5aed5c911a4fa1/src/tests/BuildWasmApps/Wasm.Build.Tests/BuildTestBase.cs#L228

I also wonder what would happen to web app, which has `dotnet.js` on path relative to `window.location`
I think we don't have unit test for that either.
(Note that this loader is not used by Blazor, they have their own)

yamachu commented:
Thanks for the review!

I have set up an environment to run the tests you presented, so I rewrote the working directory and ran the tests.
But, the working directory in run-v8.sh has changed and v8 is having problems finding test-main.js.
There is also a module resolution issue with js, so it appears that a drastic rewrite of the test environment is needed...

yamachu commented:
I have not yet been to verify this for the web app.
Looking at the emscripten implementation, it starts from the URL (window.location.href) where the execution took place.
This is very different from the behavior of path resolution in Node.

In a web environment, it may be desirable to override the locateFile, as in the following sample code.
For example, if you consider the case where you want to distribute dotnet.js or managed assets on a CDN or similar, the locateFile behavior provided by emscripten will not work.
https://github.com/dotnet/runtime/blob/main/src/mono/sample/wasm/browser-nextjs/components/deepThought.js

However, I think it is also possible to support this by utilizing `remoteSources` in addition to this locateFile change.
https://github.com/dotnet/runtime/blob/10883a7bae7fbd99b1edaa77d4ea4a5a3766114d/src/tasks/WasmAppBuilder/WasmAppBuilder.cs

Currently, only icu files support remoteSources, but I think that by making all assets support it, it will be possible to execute in a way that does not depend on the behavior of the locateFile.

yamachu commented:
## Summary

Running outside the AppBundle directory

| Patch ([a1c3c7a](https://github.com/dotnet/runtime/pull/69441/commits/a1c3c7a3651f14266b9f4bf48e9d81775303de7b)) | NodeJS | V8 | Chrome |
| - | - | - | - |
| before | Fail | Fail | Fail |
| after | Success | Needs workaround (1) | Fail |

workaround (1)

When running in a V8 (shell?) environment, `scriptDirectory` is always an empty string, so it is necessary to override locateFile and explicitly provide a base path.

test-main.js

```diff
- loadDotnet("./dotnet.js").then((createDotnetRuntime) => {
+ loadDotnet("./AppBundle/dotnet.js").then((createDotnetRuntime) => {
    return createDotnetRuntime(({ MONO, INTERNAL, BINDING, Module }) => ({
        disableDotnet6Compatibility: true,
        config: null,
        configSrc: processedArguments.run_deep_working_dir ? "./AppBundle/mono-config.json" : "./mono-config.json",
+        // @ts-ignore
+        locateFile: (path, _scriptDirectory /* V8 always '', so not needed */) => {
+            // return _scriptDirectory + path;
+            return TO_ASSETS_PATH_FROM_CWD + path;
+        },
```

Because of the above problem, I have not added V8 to the test pattern.

## What does this PR solve

1. Allows execution outside the AppBundle directory in a Node.JS environment
2. Consistent path resolution for resources enumerated in mono-config.json

pavelsavara commented on src/mono/wasm/test-main.js:
```diff
@@ -143,7 +143,7 @@ loadDotnet("./dotnet.js").then((createDotnetRuntime) => {
         disableDotnet6Compatibility: true,
         config: null,
-        configSrc: "./mono-config.json",
+        configSrc: processedArguments.run_deep_working_dir ? "./AppBundle/mono-config.json" : "./mono-config.json",
```
Perhaps 
```
await mono_wasm_load_config(locateFile(moduleExt.configSrc));
```
In  `mono_wasm_pre_init` of startup.ts would be better ?

yamachu replied:
That sounds like a good change.
However, if we adopt that modification, I think it will be difficult to determine from where the relative path in the jsdoc comment in `configFilePath` is relative.

https://github.com/dotnet/runtime/blob/97b29919e4bb7eaebf4c5cdb6b53bce3b459a0c3/src/mono/wasm/runtime/startup.ts#L668

What do you think of the idea of adding a statement that the path is relative to dotnet.js?

pavelsavara replied:
@radical I know you are thinking about redesign of the whole area. Could you please add your thoughts ?

pavelsavara replied:
I'm not looking for quick hack,  but rather approach, which would be easy to understand and which would be consistent on all targets.

pavelsavara replied:
I'm thinking about naïve approach from user perspective:
- `configSrc` is relative to `dotnet.js` unless I use absolute path
- `config.assembly_root` is relative to `configSrc` unless I use absolute path
- `asset.name` is relative to `configSrc` unless I use absolute path
- `sourcePrefix` is used only for relative paths.

Does this make sense ? Or there is better alternative ?
I admit that I didn't explore all consequences now. Is any of that breaking change ?

Why `scriptDirectory` didn't work on V8 ? 

I know we had trouble resolving it on ES6. See
https://github.com/dotnet/runtime/blob/bfbb78354e536ac616f2f9dabf3db2b8fa8b9f64/src/mono/wasm/runtime/es6/dotnet.es6.lib.js#L24
It only happen asynchronously. But that should be fine.

New approach has to work for ES6 too. 

And ideally for webpack too.
See https://github.com/webpack/webpack/issues/7352
And https://github.com/webpack/webpack/pull/15246

pavelsavara replied:
On `V8` even if the `scriptDirectory` is always empty, could we treat it as `.` and do relative paths with it ? 
This worked for me d8> `read("../xx.md")`. 

Which is what our `fetch_like` does.  See https://github.com/dotnet/runtime/blob/973b7640aed5f6f1aa95bea10361b7792f9c1c5d/src/mono/wasm/runtime/polyfills.ts#L32

yamachu replied:
That makes sense.

Here is an example of breaking changes in Node.js environment.

Current main

- `configSrc` is relative to `cwd` unless I use absolute path
- `config.assembly_root` is relative to `dotnet.js` unless I use absolute path
- `asset.name` (assembly and resource) is relative to `dotnet.js` unless I use absolute path
  - `asset.name` (other) is relative to `cwd` unless I use absolute path
- `sourcePrefix` is used only for relative paths. If it's absolute, we do not use assembly_root.

Patch https://github.com/dotnet/runtime/pull/69441/commits/a1c3c7a3651f14266b9f4bf48e9d81775303de7b

- `configSrc` is relative to `cwd` unless I use absolute path
- `config.assembly_root` is relative to `dotnet.js` unless I use absolute path
- `asset.name` (assembly and resource) is relative to `dotnet.js` unless I use absolute path
  - `asset.name` (other) is relative to **`dotnet.js`** unless I use absolute path <- changed, this seems reasonable to me.
- `sourcePrefix` is used only for relative paths. If it's absolute, we do not use assembly_root.

And https://github.com/dotnet/runtime/pull/69441/commits/838852557c4988b5dcae2e639390c3ef5e3dc439

- `configSrc` is relative to **`dotnet.js`** unless I use absolute path <- changed
- `config.assembly_root` is relative to `dotnet.js` unless I use absolute path
- `asset.name`  is relative to `dotnet.js` unless I use absolute path
- `sourcePrefix` is used only for relative paths. If it's absolute, we do not use assembly_root.

Thus, the path to some `cwd` is changed to the path to `dotnet.js`.

On V8, everything relative to `cwd` except `sourcePrefix`.

The reason can be led from several things.
When executed with the following command, the variables indicating the environment are as follows.

```sh
$ v8 test-main.js
```

- ENVIRONMENT_IS_WEB == false
- ENVIRONMENT_IS_WORKER == false
- ENVIRONMENT_IS_NODE == false
- ENVIRONMENT_IS_SHELL == true

Looking at the emscripten code, there is no value set for `scriptDirectory` in the case of `ENVIRONMENT_IS_SHELL`.

https://github.com/emscripten-core/emscripten/blob/e1d97f55d12ed63dc30e547c818af9a63be0b976/src/shell.js#L155

So the behavior of `locateFile` is simply to return the filename, which is resolved by the path to the cwd.
Even if `./` is given at the beginning of the file and treated as a relative path, it will be the path to the cwd.

However, when `dotnet.js` is loaded with `import` statement, `import.meta.url` can be retrieved, so I am thinking that it may be possible to override the `scriptDirectory` by combining it with the following code. (like https://github.com/dotnet/runtime/blob/bfbb78354e536ac616f2f9dabf3db2b8fa8b9f64/src/mono/wasm/runtime/es6/dotnet.es6.lib.js#L24)

```js
if (ENVIRONMENT_IS_SHELL)  {
  const tmp = import.meta.url;
  scriptDirectory = tmp.substring(0, tmp.lastIndexOf('/') + 1) + '/';
}
```

The combination of CommonJS and V8 would be the path to cwd unless an external base path is given...

By the way, I have not been able to research the web (Chrome).
This is because the following function fails to load `dotnet.js` and `wasm`.

https://github.com/dotnet/runtime/blob/a1c3c7a3651f14266b9f4bf48e9d81775303de7b/src/mono/wasm/runtime/startup.ts#L28

This is because the code below specifies the `href` as `dotnet.js` and uses `location.href `as the starting point unless the `base` tag is used.
https://github.com/dotnet/runtime/blob/a1c3c7a3651f14266b9f4bf48e9d81775303de7b/src/mono/wasm/runtime/startup.ts#L46

Also, the behavior through the bundler (like webpack, rollup...) has not been investigated.

yamachu replied:
I have investigated whether this change would work under ES6 environment.

In conclusion, it failed regardless of this PR change.
We need to change the Assertion in the following places to test under ES6 environment.

https://github.com/dotnet/runtime/blob/cca6bb6be279ee92325148b8525ece2ebdcb2070/src/tests/BuildWasmApps/Wasm.Build.Tests/BuildTestBase.cs#L539
(If WasmEnableES6 is true, re-link will occur and the Assertion at that point will fail according to the Config.)

The following changes would also be required.

https://github.com/dotnet/runtime/blob/cca6bb6be279ee92325148b8525ece2ebdcb2070/src/mono/wasm/test-main.js#L259-L261
(If test-main.js is executed as type: module, dotnet.js needs to be loaded with `import` statement.)

https://github.com/emscripten-core/emscripten/blob/af7f969c4497e642c4f7c256531f93b89a8b008a/src/preamble.js#L786
(URL is not defined in V8 Shell. It's difficult to bypass...)

https://github.com/dotnet/runtime/blob/cca6bb6be279ee92325148b8525ece2ebdcb2070/src/mono/wasm/runtime/es6/dotnet.es6.lib.js#L50-L51
(If `IS_SHELL`, `scriptDirectory` must be overwritten using `import.meta.url`)

Thus, we have a lot of interest in the changes we want to make.
If we are also going to test in an ES6 environment, I thought we needed to resolve https://github.com/dotnet/runtime/issues/63337

yamachu replied:
Since https://github.com/dotnet/runtime/pull/70746  was merged, I would consider testing again.

yamachu replied:
https://github.com/dotnet/runtime/pull/69441#discussion_r882945695

I wrote as "configSrc is relative to dotnet.js unless I use absolute path <- changed" in this reply, but it was different.
If configSrc is given as an absolute path, the current change does not resolve the configSrc path correctly...
https://github.com/dotnet/runtime/pull/69441/commits/838852557c4988b5dcae2e639390c3ef5e3dc439

https://github.com/dotnet/runtime/pull/69441/files#diff-3e3e5eb583ed48d37f0ab14c4060488a683d4088eff83a58be530cfd43fcb213R31-R54

For this change, if path is an absolute path or uri containing a protocol such as https, we need to make a change to return only path.

like this
```js
const toAbsoluteUrl = function toAbsoluteUrl(path, prefix) {
  ...
  if (isAbsoluteOrUriOr....(path)) {
    return path;
  }
}
```

Or, if it is determined that the path is not relative, it is necessary to treat it as an absolute path and fetch it instead of using `locateFile`.

https://github.com/dotnet/runtime/pull/69441/files#diff-5cdf20fb11670158ec01622e5d8e6764608b9846dfd31ac45c85bde7f0ed4228R687

yamachu replied:
https://github.com/dotnet/runtime/pull/69441#discussion_r882394336

As pointed out there, it seems to me that `config.assembly_root` , etc. should be relative to `configSrc`.
My current PR does not support this behavior because it is relative to `dotnet.js`.

As a hack, this can be achieved by replacing `Module["locateFile"]` with a function relative to `configSrc` in `onConfigLoaded`.
I have also tried wrapping the `locateFile` used in `mono_download_assets` and using `configSrc` as the origin, but I cannot determine the best way to do this...

pavelsavara replied:
In my PR it's relative to `dotnet.js`. Maybe that's good enough ? People could still put absolute path there.

yamachu replied:
I think we can start with a relative path in `dotnet.js` .
I am assuming this is for web and am wondering what happens if I use a bundler (e.g. webpack or vite).
But I think it is good to leave that out of scope at this stage.

pavelsavara replied:
There is range of issues with webpack+dotnet (emscripten). One of them is dynamic `import("something")` and another `import.meta.url`.

There is my last attempt to make it work 
https://github.com/dotnet/runtime/blob/main/src/mono/sample/wasm/browser-webpack/webpack.config.js#L9

yamachu replied:
Thanks for the suggestions on how to deal with this.
I thought dotnet.js would be a good starting point since the means seems to exist.

pavelsavara commented on src/mono/wasm/test-main.js:
```diff
@@ -158,6 +159,7 @@ function initRunArgs() {
     runArgs.enableGC = runArgs.enableGC === undefined ? true : runArgs.enableGC;
     runArgs.diagnosticTracing = runArgs.diagnosticTracing === undefined ? false : runArgs.diagnosticTracing;
     runArgs.debugging = runArgs.debugging === undefined ? false : runArgs.debugging;
+    runArgs.deepWorkDir = runArgs.deepWorkDir === undefined ? false : runArgs.deepWorkDir;
```
I think we will re-visit `runArgs` in near future, so I'm hesitant to add more to it. 
Is this just to keep the behaior backward compatible in some scenario ? 
Which scenarios would be affected if it was true by default ?

yamachu replied:
If the default value is true, the test running on the V8 engine will search deeper than the PATH where the file actually exists when reading dotnet.wasm and the test will fail.
Therefore, this will affect all tests that test with V8.

This value is only used in the RunOutOfAppBundle test, but a change in the default value would have many affects...

pavelsavara replied:
Could that be fixed only for V8 inside `dotnet.js` without making it configurable ?

yamachu replied:
If it is possible to overwrite `scriptDirectory` with `import.meta.url` before loading dotnet.wasm in https://github.com/dotnet/runtime/blob/main/src/mono/wasm/runtime/es6/dotnet.es6.pre.js , for example, you may not need to perform this hack.
However, this will not work if the emscripten initialization script overwrites it with `var scriptDirectory` or something similar.

pavelsavara replied:
Maybe we could install better default `Module.locateFile` some merge of your 
`scriptDirectory = import.meta.url.slice(0, import.meta.url.lastIndexOf('/')) + '/';` with `test-main.js` 
```
let toAbsoluteUrl = function (path, prefix) {
    if (prefix.startsWith("/")) {
        return path;
    }
    return prefix + path;
}
if (is_browser) {
    const anchorTagForAbsoluteUrlConversions = document.createElement('a');
    toAbsoluteUrl = function toAbsoluteUrl(path, prefix) {
        anchorTagForAbsoluteUrlConversions.href = prefix + path;
        return anchorTagForAbsoluteUrlConversions.href;
    }
}
```

And do it in https://github.com/dotnet/runtime/blob/6f7064dc170a46d1a5e67823f5178b8862b27a58/src/mono/wasm/runtime/exports.ts#L250

only if there is no user override

pavelsavara commented on src/mono/wasm/test-main.js:
```diff
@@ -342,6 +346,17 @@ if (typeof globalThis.crypto === 'undefined') {
 }
 
 let toAbsoluteUrl = function (path, prefix) {
+    if (is_shell) {
```
Also, would it be possible to take the logic and move it from the `*main.js ` inside `dotnet.js` ?
Also for CJS.

yamachu replied:
https://github.com/dotnet/runtime/pull/69441#discussion_r907719667
If the scriptDirectory issue indicated in this comment were resolved, it could be run in a directory other than AppBundle (I think it would still work if it were in AppBundle).

With V8 in the CJS environment, it is not possible to identify the directory where the script is located ( it might be possible, but I don't know how to do it), so porting it to CJS would be difficult.
I assume that I would need to explicitly pass the scriptDirectory to the createDotnetRuntime function, or pass a customized locateFile.

pavelsavara replied:
I think it's OK that V8 must be run in the current directory in CJS mode.

yamachu replied:
I removed `fs` and `process` module from app-support.mjs and removed the code that `throw Error` for anything other than `is_node` and it works in v8.

yamachu commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -675,6 +675,16 @@ export function mono_wasm_load_data_archive(data: Uint8Array, prefix: string): b
 
+function is_path_absolute(path: string): boolean {
+    if (ENVIRONMENT_IS_NODE || ENVIRONMENT_IS_SHELL) {
+        return path.startsWith("/") || path.indexOf(":") !== -1;
```
`:` is used assuming Windows drive letter and absolute URI

pavelsavara commented:
@yamachu 
I took some of your ideas into https://github.com/dotnet/runtime/pull/72275.
I would appreciate if you could test that it solves your scenario.

After it's merged, we could rebase this PR and use the unit test you created.

pavelsavara commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -707,10 +717,11 @@ export function mono_wasm_load_data_archive(data: Uint8Array, prefix: string): b
  */
 export async function mono_wasm_load_config(configFilePath: string): Promise<void> {
     const module = Module;
+    const resolvedConfigFilePath = is_path_absolute(configFilePath) ? configFilePath : locateFile(configFilePath);
```
I didn't do this detection in my PR. I guess we always only supported relative paths for all resources so far ? Is there strong scenario for absolute path ?

yamachu replied:
For example, when used on the web, I thought there was a use case for distributing resources such as assemblies at different origins.
However, since there are more use cases where the AppBundle directory including dotnet.wasm is distributed as a single entity, it seems unnecessary to determine the absolute path.

pavelsavara replied:
okay, this is my take on it 
https://github.com/dotnet/runtime/blob/75192bbc6f06f6471bc9d3c3e4ffbaad19020394/src/mono/wasm/runtime/imports.ts#L182

pavelsavara commented:
@yamachu The startup changes PR is merged. 
I would like to take just the `Wasm.Build.Tests` changes from this PR. 

Could you please re-base it and remove the other changes ? Alternatively I could open another PR myself if you prefer that.
Many thanks.

yamachu committed:
Rebase resolve-asset-path-by-locatefile onto main with reset runtime changes

Squashed commit of the following:

commit 8dd23f8b8510330f18771faf1c9ca26ee550176e
Merge: ffe4a2a2330 3ac16299ebe
Author: yamachu <yamachu.dev@gmail.com>
Date:   Tue Jul 12 02:13:41 2022 +0900

    Merge branch 'main' into resolve-asset-path-by-locatefile

commit ffe4a2a233010f8727876287d529754365cb9f9f
Merge: 357ca535049 32d0360b73b
Author: yamachu <yamachu.dev@gmail.com>
Date:   Tue Jul 5 01:50:45 2022 +0900

    Merge branch 'main' into resolve-asset-path-by-locatefile

commit 357ca53504982b0719ecce39af75348055fe7a5e
Author: yamachu <yamachu.dev@gmail.com>
Date:   Fri Jul 1 02:24:48 2022 +0900

    If an absolute path is passed to configSrc, do not use locateFile to resolve the path, but fetch the path as is.

commit cf45c0d44a943ce497c9d162c871136adf391501
Merge: 527ad985b35 85223ce6cb3
Author: yamachu <yamachu.dev@gmail.com>
Date:   Fri Jul 1 01:16:11 2022 +0900

    Merge branch 'main' into resolve-asset-path-by-locatefile

commit 527ad985b350b94744c852da6b903a78c4ea72c1
Author: yamachu <yamachu.dev@gmail.com>
Date:   Tue Jun 28 16:58:28 2022 +0900

    Remove toAbsoluteUrl from test code

commit c339a1a0a9aacab48affe787a606d6ab106aca1a
Author: yamachu <yamachu.dev@gmail.com>
Date:   Tue Jun 28 16:58:10 2022 +0900

    Ported toAbsoluteUrl in the web environment to pre-module

commit 3b4f3f19ad9c29cc20d4ddb7a06115297f69befb
Author: yamachu <yamachu.dev@gmail.com>
Date:   Tue Jun 28 14:18:16 2022 +0900

    Remove unused variable

commit 2cdd2c7c9e0d7df8641f88be70b517a31e8a5634
Author: yamachu <yamachu.dev@gmail.com>
Date:   Tue Jun 28 14:14:54 2022 +0900

    Unassign toAbsoluteUrl to use locateFile to correct script directory in Node and V8 environments

commit e7a473afcf664a0358679d8ecb7777364ffdd8d6
Author: yamachu <yamachu.dev@gmail.com>
Date:   Tue Jun 28 14:13:05 2022 +0900

    Assign a locateFile to the Module that corrects the prefix

    Fix that prefix argument does not indicate the location where the script exists when dotnet.wasm is loaded

commit d54b1436addd5e074ccdec6b94b4c36303c54c30
Author: yamachu <yamachu.dev@gmail.com>
Date:   Tue Jun 28 14:05:06 2022 +0900

    Remove flags
     that is only used in certain scenarios

commit 71c1ff79e1d5e1fe7b58c2515404a85ef31c6386
Author: yamachu <yamachu.dev@gmail.com>
Date:   Tue Jun 28 03:35:26 2022 +0900

    Refactor: use already implemented attributes

commit 86227f1ce580fbff3ccce9c499e6db45d4d84b4b
Author: yamachu <yamachu.dev@gmail.com>
Date:   Tue Jun 28 02:03:20 2022 +0900

    Add notes

commit bac891e143cdaadd6b0f1befb3a3003c9f39f758
Author: yamachu <yamachu.dev@gmail.com>
Date:   Tue Jun 28 00:30:42 2022 +0900

    test all RunHost

    The project format changes depending on the host, so projects with the same id cannot be used interchangeably.
    Therefore, the project is re-created so that the test id changes, and the cache is also discarded.

commit f41fd42255c81a113ba61ed3de6976ef0bcd6e6b
Author: yamachu <yamachu.dev@gmail.com>
Date:   Tue Jun 28 00:29:53 2022 +0900

    Modified runtime to get the path starting from dotnet.js in shell environment and use that path in test

commit 3e2476cf6250637c8601d0fe112549f363b353d9
Author: yamachu <yamachu.dev@gmail.com>
Date:   Tue Jun 28 00:25:02 2022 +0900

    "--run" prefix is reserved, so change flag name

commit ee820c5623a44f87c1095a747bfcb652fa76bacf
Merge: 30768018631 c9f80fdde29
Author: yamachu <yamachu.dev@gmail.com>
Date:   Mon Jun 27 09:43:29 2022 +0900

    Merge branch 'main' into resolve-asset-path-by-locatefile

commit 30768018631636ef5978087121c45b45f85c88a5
Merge: 838852557c4 cca6bb6be27
Author: yamachu <yamachu.dev@gmail.com>
Date:   Thu Jun 9 21:38:08 2022 +0900

    Merge branch 'main' into resolve-asset-path-by-locatefile

commit 838852557c4988b5dcae2e639390c3ef5e3dc439
Author: yamachu <yamachu.dev@gmail.com>
Date:   Thu May 26 03:34:56 2022 +0900

    Also use locateFile to resolve configFile paths

commit 97b29919e4bb7eaebf4c5cdb6b53bce3b459a0c3
Author: yamachu <yamachu.dev@gmail.com>
Date:   Sat May 21 22:47:57 2022 +0900

    Add Unit test to run WasmApp outside of AppBundle directory

    V8: Could not determine scriptDirectory (always empty string). You must explicitly specify the base path within locateFile.
    Chrome: Failed to load dotnet.wasm

commit a1c3c7a3651f14266b9f4bf48e9d81775303de7b
Author: yamachu <yamachu.dev@gmail.com>
Date:   Wed May 18 00:57:26 2022 +0900

    Use localeFile to resolve path for assets outside of remoteSources

    Use localeFile to resolve path for assets outside of remoteSources, so that we can run from a different path than locate that dotnet.js exists

    ref:  #69259

yamachu commented:
The merge commit was in a branch, which made rebase difficult.
Therefore, I generated a squash commit to deal with the problem.

I'm sorry for not being able to follow up on your review.

I removed changes related to runtime and picked changes related to test.

I've tested it on local and it passes all the scenarios I had in mind!
Thanks for the great improvement.

pavelsavara commented on src/mono/wasm/test-main.js:
```diff
@@ -217,6 +218,12 @@ function processQueryArguments(incomingArguments) {
+        } else if (currentArg.startsWith("--config-src=")) {
+            const arg = currentArg.substring("--config-src=".length);
+            runArgs.configSrc = arg;
+        } else if (currentArg == ("--deep-work-dir")) {
```
we could remove the empty condition

yamachu replied:
If we remove that, it will go into the `else` clause and the while loop will `break`.
Unhandled arguments will then be passed as `applicationArgument` , which may cause unintended behavior.

pavelsavara commented on src/mono/wasm/test-main.js:
```diff
@@ -355,7 +362,7 @@ Promise.all([argsPromise, loadDotnetPromise]).then(async ([_, createDotnetRuntim
         disableDotnet6Compatibility: true,
         config: null,
-        configSrc: "./mono-config.json",
+        configSrc: runArgs.configSrc,
```
`configSrc: runArgs.configSrc || "./mono-config.json",`

pavelsavara commented on src/tests/BuildWasmApps/Wasm.Build.Tests/WasmRunOutOfAppBundleTests.cs:
```diff
@@ -0,0 +1,60 @@
+                }
+            }
+
+            RunAndTestWasmApp(buildArgs, expectedExitCode: 42, host: host, id: id, extraXHarnessMonoArgs: "--deep-work-dir");
```
do we need `--deep-work-dir` here ?

pavelsavara commented:
I updated the title and the description of the PR.
@radical could you please help @yamachu to finish their contribution.

yamachu committed:
apply suggestions that assign configSrc

pavelsavara commented:
/azp run runtime-wasm

radical commented on src/tests/BuildWasmApps/Wasm.Build.Tests/WasmRunOutOfAppBundleTests.cs:
```diff
@@ -0,0 +1,60 @@
+        {}
+
+        [Theory]
+        [BuildAndRun(aot: false, host: RunHost.V8)]
+        [BuildAndRun(aot: true, host: RunHost.V8)]
+        [BuildAndRun(aot: false, host: RunHost.Chrome)]
+        [BuildAndRun(aot: true, host: RunHost.Chrome)]
+        [BuildAndRun(aot: false, host: RunHost.NodeJS)]
+        [BuildAndRun(aot: true, host: RunHost.NodeJS)]
```
You can use `[BuildAndRun]`, and it will add all these combinations by default.

yamachu replied:
If we replace it with `[BuildAndRun]`, we will get an error like this.
`InitProjectDir` will fail because the test is performed on the same id.

https://github.com/dotnet/runtime/blob/2db797485e4a885aae7572806090c42accc2c62f/src/tests/BuildWasmApps/Wasm.Build.Tests/BuildTestBase.cs#L281

```
  ==============================================================================================
  =============== Running with Workloads ===============
  ==============================================================================================

      Wasm.Build.Tests.WasmRunOutOfAppBundleTests.RunOutOfAppBundle(buildArgs: BuildArgs { ProjectName = placeholder, Config = Debug, AOT = False, ProjectFileContents = placeholder, ExtraBuildArgs =  }, host: V8, id: "2kfi50gc.hjx") [STARTING]
      Wasm.Build.Tests.ConfigSrcTests.ConfigSrcAbsolutePath(buildArgs: BuildArgs { ProjectName = placeholder, Config = Debug, AOT = False, ProjectFileContents = placeholder, ExtraBuildArgs =  }, host: V8, id: "jnopof1n.ffi") [STARTING]
      Wasm.Build.Tests.WasmRunOutOfAppBundleTests.RunOutOfAppBundle(buildArgs: BuildArgs { ProjectName = placeholder, Config = Debug, AOT = False, ProjectFileContents = placeholder, ExtraBuildArgs =  }, host: V8, id: "2kfi50gc.hjx") [FINISHED] Time: 8.4030997s
      Wasm.Build.Tests.WasmRunOutOfAppBundleTests.RunOutOfAppBundle(buildArgs: BuildArgs { ProjectName = placeholder, Config = Debug, AOT = False, ProjectFileContents = placeholder, ExtraBuildArgs =  }, host: Chrome, id: "2kfi50gc.hjx") [STARTING]
      Wasm.Build.Tests.ConfigSrcTests.ConfigSrcAbsolutePath(buildArgs: BuildArgs { ProjectName = placeholder, Config = Debug, AOT = False, ProjectFileContents = placeholder, ExtraBuildArgs =  }, host: V8, id: "jnopof1n.ffi") [FINISHED] Time: 8.4034685s
      Wasm.Build.Tests.ConfigSrcTests.ConfigSrcAbsolutePath(buildArgs: BuildArgs { ProjectName = placeholder, Config = Debug, AOT = False, ProjectFileContents = placeholder, ExtraBuildArgs =  }, host: NodeJS, id: "jnopof1n.ffi") [STARTING]
      Wasm.Build.Tests.WasmRunOutOfAppBundleTests.RunOutOfAppBundle(buildArgs: BuildArgs { ProjectName = placeholder, Config = Debug, AOT = False, ProjectFileContents = placeholder, ExtraBuildArgs =  }, host: Chrome, id: "2kfi50gc.hjx") [FAIL]
        System.IO.IOException : The file '/Users/yk-yamada/Projects/github.com/yamachu/dotnet-runtime/artifacts/bin/Wasm.Build.Tests/Release/net7.0/browser-wasm/2kfi50gc.hjx/nuget.config' already exists.
        Stack Trace:
             at Interop.ThrowExceptionForIoErrno(ErrorInfo errorInfo, String path, Boolean isDirectory)
             at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String path, OpenFlags flags, Int32 mode, Func`4 createOpenException)
             at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize, UnixFileMode openPermissions, Int64& fileLength, UnixFileMode& filePermissions, Func`4 createOpenException)
             at System.IO.FileSystem.CopyFile(String sourceFullPath, String destFullPath, Boolean overwrite)
          /Users/yk-yamada/Projects/github.com/yamachu/dotnet-runtime/src/tests/BuildWasmApps/Wasm.Build.Tests/BuildTestBase.cs(289,0): at Wasm.Build.Tests.BuildTestBase.InitProjectDir(String dir)
          /Users/yk-yamada/Projects/github.com/yamachu/dotnet-runtime/src/tests/BuildWasmApps/Wasm.Build.Tests/BuildTestBase.cs(343,0): at Wasm.Build.Tests.BuildTestBase.BuildProject(BuildArgs buildArgs, String id, BuildProjectOptions options)
          /Users/yk-yamada/Projects/github.com/yamachu/dotnet-runtime/src/tests/BuildWasmApps/Wasm.Build.Tests/WasmRunOutOfAppBundleTests.cs(24,0): at Wasm.Build.Tests.WasmRunOutOfAppBundleTests.RunOutOfAppBundle(BuildArgs buildArgs, RunHost host, String id)
             at InvokeStub_WasmRunOutOfAppBundleTests.RunOutOfAppBundle(Object, Object, IntPtr*)
             at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
      Wasm.Build.Tests.WasmRunOutOfAppBundleTests.RunOutOfAppBundle(buildArgs: BuildArgs { ProjectName = placeholder, Config = Debug, AOT = False, ProjectFileContents = placeholder, ExtraBuildArgs =  }, host: Chrome, id: "2kfi50gc.hjx") [FINISHED] Time: 0.0153179s
```

radical commented on src/tests/BuildWasmApps/Wasm.Build.Tests/ConfigSrcTests.cs:
```diff
@@ -0,0 +1,41 @@
+
+        // NOTE: port number determinizes dynamically, so could not generate absolute URI
+        [Theory]
+        [BuildAndRun(aot: false, host: RunHost.V8)]
+        [BuildAndRun(aot: true, host: RunHost.V8)]
+        [BuildAndRun(aot: false, host: RunHost.NodeJS)]
+        [BuildAndRun(aot: true, host: RunHost.NodeJS)]
```
This should work I think - `[BuildAndRun(host: RunHost.V8 | RunHost.NodeJS)]`

radical commented:
You need to add the new test classes to https://github.com/dotnet/runtime/blob/main/eng/testing/scenarios/BuildWasmAppsJobsList.txt .

yamachu committed:
apply suggestions

radical commented on src/tests/BuildWasmApps/Wasm.Build.Tests/WasmRunOutOfAppBundleTests.cs:
```diff
@@ -0,0 +1,60 @@
+                            new BuildProjectOptions(
+                                InitProject: () => File.WriteAllText(Path.Combine(_projectDir!, "Program.cs"), s_mainReturns42),
+                                DotnetWasmFromRuntimePack: !(buildArgs.AOT || buildArgs.Config == "Release"),
+                                UseCache: false));
```
Try removing this. Basically, what we do is build once for a config+aot-setting, and then use the same build to run with different hosts. So, it needs the cache for that.

yamachu replied:
When using cache, the `AppBundle` directory was generated in a single nested directory.

radical replied:
Would the test still be valid if you were to copy the `AppBundle` to a new temporary directory, and try to run from there? that would simulate running from a different directory too, right?

And what does `--deep-work-dir` do? It seems to be a nop currently.

yamachu replied:
You’re right.
It looks like the test can be done by copying it from the AppBundle directory to another temporary directory and running it.

The `--deep-work-dir` is used to change the entry point in BuildTestBase for the execution target.
https://github.com/dotnet/runtime/pull/69441#discussion_r930099873

radical replied:
Instead of `--deep-work-dir` stuff, add a new param - `jsRelativePath` defaulting to `test-main.js`, to `RunAndTestWasmApp`.

yamachu committed:
Add tests to BuildWasmAppsJobs

yamachu committed:
add jsRelativePath args instead of using extraXHarnessMonoArgs hack

yamachu committed:
Copy AppBundle directory and treat it as the test target.

radical commented on src/tests/BuildWasmApps/Wasm.Build.Tests/ConfigSrcTests.cs:
```diff
@@ -0,0 +1,38 @@
+
+#nullable enable
+
+namespace Wasm.Build.Tests
```
nit: 
```suggestion
namespace Wasm.Build.Tests;
```

yamachu replied:
Is it meant to add a semicolon, remove the bracket, and lower the indentation level by one?

like this
```cs
namespace Wasm.Build.Tests;

public class WasmRunOutOfAppBundleTests : BuildTestBase
{
    // Impl
}
```

radical replied:
Yes, file scoped namespaces. Sorry, I should have been clearer.

radical commented on src/tests/BuildWasmApps/Wasm.Build.Tests/WasmRunOutOfAppBundleTests.cs:
```diff
@@ -0,0 +1,83 @@
+
+#nullable enable
+
+namespace Wasm.Build.Tests
```
nit:
```suggestion
namespace Wasm.Build.Tests;
```

radical commented on src/tests/BuildWasmApps/Wasm.Build.Tests/WasmRunOutOfAppBundleTests.cs:
```diff
@@ -0,0 +1,83 @@
+            string binDir = GetBinDir(baseDir: _projectDir!, config: buildArgs.Config);
+            string appBundleDir = Path.Combine(binDir, "AppBundle");
+            string tmpBundleDirName = "AppBundleTmp";
+            string tmpBundleDir = Path.Combine(binDir, tmpBundleDirName);
```
What do you think of creating this tmpBundleDir in a completely different location, like under `Path.GetTempPath()`? This would ensure that it isn't dependent on anything in the binDir.
And make sure to add `id`, and the project name to the dir name, so the different runs don't overlap.

yamachu replied:
It is a good method in Node and V8 environments!
I' ll change that style.

However, browsers like Chrome cannot explore the parent directory, so I would like to keep the current style.

yamachu committed:
convert to file scoped namespace style

yamachu committed:
use tmp directory to avoid running overlap

radical commented on src/tests/BuildWasmApps/Wasm.Build.Tests/WasmRunOutOfAppBundleTests.cs:
```diff
@@ -0,0 +1,84 @@
+                File.WriteAllText(indexHtmlPath, html);
+            }
+        } else {
+            CopyAllFiles(appBundleDir, tmpDir);
```
You can use `Utils.DirectoryCopy(string sourceDirName, string destDirName, Func<string, bool>? predicate=null, bool copySubDirs=true, bool silent=false)`

yamachu committed:
use Utils.DirectoryCopy

radical commented:
Don't merge yet. The test needs some changes. I will get those working locally, and share that tomorrow.

radical commented:
So, one of the problems with the `WasmRunOutOfAppBundle` test is that we won't be able to run `dotnet xharness` outside in a temporary directory. That can be worked around by creating a tool manifest there, but then we have to get the xharness version from the repo to helix. This applies to all the hosts.

IIUC, the intention here is:
1. `v8`, `node` - start from a directory which is *not* the `AppBundle` directory. In which case we can use the parent directory.
2. `chrome` - the web server can be started in the parent directory too.

This is what I came up with: https://gist.github.com/radical/18d60949c789cf98f8b175c30096b683 . Feel free to add this to your PR.

The test output has:
- `Running v8 --expose_wasm ./AppBundle/test-main.js ...`
- `Running node ./AppBundle/test-main.js ...`

Make sure to confirm that it is indeed testing what is intended here. AFAIU, it does that.

yamachu committed:
apply patch that allow configuring bundleDir

see: https://github.com/dotnet/runtime/pull/69441#issuecomment-1196401080

radical commented:
@yamachu Thank you for your contribution! Great job 👍

yamachu commented:
@pavelsavara @radical 
Thanks for all your support!

radical commented:
Failures are unrelated.

radical commented:
@yamachu This is only adding the `--config-src` param to `test-main.js`, which is not used by any user apps. Was that the intention here? I don't see the reason for adding that, since tests run in a pretty controlled environment.

yamachu commented:
This addition it to verify that config-src could be resolved when passed as an absolute path.
ref: https://github.com/dotnet/runtime/pull/69441#discussion_r910258202
